python/lsst/daf/butler/_butler.py

```diff
@@ -111,6 +114,19 @@
 log = getLogger(__name__)
 
 
+class DatasetExistence(Flag):
+    Unknown = 0
+    RegistryKnows = auto()
+    DatastoreKnows = auto()
+    ArtifactExists = auto()
+    ArtifactNotChecked = auto()
+    Exists = RegistryKnows | DatastoreKnows | ArtifactExists
+    PartialCheck = RegistryKnows | DatastoreKnows | ArtifactNotChecked
```

Comment:
I'd like to handle `ArtifactNotChecked` differently somehow; we also don't check artifact existence when `~RegistryKnows` or `~DatastoreKnows`, so we should at least rename it to something like `ArtifactAssumed`.  But I also think it might be better to make `DatastoreKnows & ~ArtifactDoesNotExist` back into an error, since that's really not a state the repository should ever be in (and that'll be even more seriously the case after DMTN-249).  We could then drop the `ArtifactNotChecked` flag, with the default being that artifact existence being assumed if `DatastoreKnows` and verified if `ArtifactExists`.  We could also consider dropping `ArtifactExists`, too, since its value would then be wholly deterministic from `DatastoreKnows & full_check`, and the caller could be expected to know what they passed as `full_check`, but I have no objection to keeping it.

If we keep `ArtifactExists`, I think we should also rename `PartialCheck` (` = RegistryKnows | DatastoreKnows`) to something like `FastExists` and rename `Exists` to `ExhaustiveExists` so the most default "does exist" result doesn't imply that the user should have done more.

I think we could accomplish all that in terms of flags, but here's a regular enum version that may be worth considering:
```
class DatasetExistence(Enum):
    UNRECOGNIZED = auto()   # No matching dataset in Registry.
    RECORDED = auto()   # In Registry, but not in any Datastore (per datastore records)
    COMPLETE = auto()  # In Registry and Datastore
    EXHAUSTIVE = auto()  # In Registry and Datastore, artifact verified

    def __bool__(self) -> bool:
        return self is self.COMPLETE or self is self.EXHAUSTIVE
```
The advantage of this is that it can only represent states that make sense (e.g. no `~RegistryKnows & DatastoreKnows` or `ArtifactExists & ArtifactNotChecked`).  The disadvantage is that you can't just OR in new flags to "upgrade" the state.

Reply:
I used `ArtifactNotChecked` because it seemed to simplify the bool logic (and I had to check `.value` because otherwise I ended up with recursion errors) and allowed the "truth" checks to work consistently -- I wasn't expecting it to be a public flag that people would check against but was an internal assertion that everything was okay and had been checked.

Reply:
I also wasn't expecting PartialCheck to be public. It was also there to simplify the boolean. I was assuming that people would either just use truthiness or be asking for RegistryKnows or DatastoreKnows explicitly.

Reply:
My other comment is that I much prefer compound enum states. RECORDED meaning in registry but not in datastore makes it harder to code because it means that you need to keep track of the registry state and then later decide whether to override with datastore state. Whereas OR-ing together as you go and then using combination flags for the public information is easier to calculate at the end.

Reply:
I've rewritten and renamed the Flags and hidden flag names for the artifact existence work well.